### PR TITLE
Create models, migrations, helpers and unit test for new tables

### PR DESCRIPTION
**Pull request by Beckyrose200**
https://eaflood.atlassian.net/browse/WATER-4188

This PR is focusing on creating our 3 new tables for the 2pt matching. These are `review_charge_element_results`, `review_return_results` and `review_results`. This PR includes the migrations, models, helpers and unit tests.

### CODE REVIEW

**Comment by Beckyrose200** [comment]
* Need to check the purposes column on the `review_return_results` table helper

File: test/models/review-charge-element-result.model.test.js
Comment (Cruikshanks) [inline test/models/review-charge-element-result.model.test.js:1]: The `ReviewChargeElementResultModel` has a relationship to `ReviewResult` which isn't covered here. It's something we normally add and I think we should continue to do so.

File: test/models/review-result.model.test.js [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+          .findById(testRecord.id)
+          .withGraphFetched('reviewChargeElementResults')
+
+        expect(result).to.be.instanceOf(ReviewResultModel)
+        expect(result.id).to.equal(testRecord.id)
+
+        expect(result.reviewChargeElementResults[0]).to.be.instanceOf(ReviewChargeElementResultModel)
+        expect(result.reviewChargeElementResults[0]).to.equal(testReviewChargeElementResult)
+      })
+    })
```
Comment (Cruikshanks) [inline test/models/review-result.model.test.js:68]: If you look at tests for other models we have 2 test patterns we follow; one for `BelongsToOneRelation` and another for `HasManyRelation`.

The key difference is whether we create 1 or multiple test records, and then confirm if we get a single object or array of them back.

This test is the pattern for `BelongsToOneRelation` but the relationship is actually a `HasManyRelation`.

If it helps `test/models/crm-v2/billing-account.model.test.js` has an example of both types that doesn't get too mad!

File: test/models/review-result.model.test.js [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+          .withGraphFetched('reviewReturnResults')
+
+        expect(result).to.be.instanceOf(ReviewResultModel)
+        expect(result.id).to.equal(testRecord.id)
+
+        expect(result.reviewReturnResults[0]).to.be.instanceOf(ReviewReturnResultModel)
+        expect(result.reviewReturnResults[0]).to.equal(testReviewReturnResult)
+      })
+    })
+  })
```
Comment (Cruikshanks) [inline test/models/review-result.model.test.js:98]: The previous comment applies here as well.

File: test/models/review-return-result.model.test.js
Comment (Cruikshanks) [inline test/models/review-return-result.model.test.js:1]: The `ReviewReturnResultModel` has a relationship to `ReviewResult` which isn't covered here. It's something we normally add and I think we should continue to do so.

File: test/support/helpers/review-charge-element-result.helper.js [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+/**
+ * Add a new charge element result for 2pt matching
+ *
+ * If no `data` is provided, default values will be used. These are
+ *
+ * - `id` - [random UUID]
+ * - `chargeElementId` - [random UUID]
+ * - allocated - 0
+ * - aggregate - 1
+ * - chargeDatesOverlap - false
```
Comment (Cruikshanks) [inline test/support/helpers/review-charge-element-result.helper.js:19]:
```suggestion
 * - `id` - [random UUID]
 * - `chargeElementId` - [random UUID]
 * - `allocated` - 0
 * - `aggregate` - 1
 * - `chargeDatesOverlap` - false
```

File: test/support/helpers/review-charge-element-result.helper.js [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+ * - updatedAt - new Date()
```
Comment (Cruikshanks) [inline test/support/helpers/review-charge-element-result.helper.js:20]: The `created_at` and `updated_at` get automatically populated when the record is created. So, we shouldn't need to default them.

File: test/support/helpers/review-result.helper.js [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+ * If no `data` is provided, default values will be used. These are
+ *
+ * - `id` - [random UUID]
+ * - `billRunId` - [random UUID]
+ * - `licenceId` - [random UUID]
+ * - `chargeVersionId` - [random UUID]
+ * - `chargeReferenceId` - [random UUID]
+ * - chargePeriodStartDate - 2022-04-01
+ * - chargePeriodEndDate - 2022-06-05
+ * - chargeVersionChangeReason - 'Strategic review of charges (SRoC)'
```
Comment (Cruikshanks) [inline test/support/helpers/review-result.helper.js:22]:
```suggestion
 * - `chargePeriodStartDate` - 2022-04-01
 * - `chargePeriodEndDate` - 2022-06-05
 * - `chargeVersionChangeReason` - Strategic review of charges (SRoC)
```

File: test/support/helpers/review-result.helper.js [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+ * - `reviewChargeElementResultId` - [random UUID]
+ * - `reviewReturnResultId` - [random UUID]
+ * - updatedAt - new Date()
```
Comment (Cruikshanks) [inline test/support/helpers/review-result.helper.js:25]: The `created_at` and `updated_at` get automatically populated when the record is created. So, we shouldn't need to default them.

File: test/support/helpers/review-return-result.helper.js [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+ * - dueDate - 2022-06-03
+ * - receivedDate - 2022-06-03
+ * - status - `completed`
+ * - underQuery - false
+ * - nilReturn - false
+ * - description - `Lands at Mosshayne Farm, Exeter & Broadclyst`
+ * - purposes - {}
+ * - quantity - 0
+ * - allocated - 0
+ * - abstractionOutsidePeriod - false
```
Comment (Cruikshanks) [inline test/support/helpers/review-return-result.helper.js:32]:
```suggestion
 * - `startDate` - 2022-04-01
 * - `endDate` - 2022-05-06
 * - `dueDate` - 2022-06-03
 * - `receivedDate` - 2022-06-03
 * - `status` - completed
 * - `underQuery` - false
 * - `nilReturn` - false
 * - `description` - Lands at Mosshayne Farm, Exeter & Broadclyst
 * - `purposes` - {}
 * - `quantity` - 0
 * - `allocated` - 0
 * - `abstractionOutsidePeriod` - false
```

File: test/support/helpers/review-return-result.helper.js [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+ * - updatedAt - new Date()
+ *
```
Comment (Cruikshanks) [inline test/support/helpers/review-return-result.helper.js:34]: The `created_at` and `updated_at` get automatically populated when the record is created. So, we shouldn't need to default them.

**Review by Cruikshanks — CHANGES_REQUESTED** [review]
Few things I think we need to update. The key thing is revisiting the unit tests, either to add some additional ones or switch the pattern being used.

I hope it makes sense.

File: test/models/review-result.model.test.js [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+    it('can successfully run a basic query', async () => {
+      const result = await ReviewResultModel.query().findById(testRecord.id)
+
+      expect(result).to.be.an.instanceOf(ReviewResultModel)
+      expect(result.id).to.equal(testRecord.id)
+    })
+  })
+
+  describe('Relationships', () => {
+    describe('when linking to review charge element result', () => {
```
Comment (Cruikshanks) [inline test/models/review-result.model.test.js:42]:
```suggestion
    describe('when linking to review charge element results', () => {
```

File: test/models/review-result.model.test.js [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+        expect(result[1].id).to.equal(testRecords[1].id)
+
+        expect(result[0].reviewChargeElementResults).to.be.an.array()
+        expect(result[1].reviewChargeElementResults).to.be.an.array()
+
+        expect(result[0].reviewChargeElementResults[0]).to.be.instanceOf(ReviewChargeElementResultModel)
+        expect(result[1].reviewChargeElementResults[0]).to.be.instanceOf(ReviewChargeElementResultModel)
+
+        expect(result[0].reviewChargeElementResults).to.include(testReviewChargeElementResult)
+        expect(result[1].reviewChargeElementResults).to.include(testReviewChargeElementResult)
```
Comment (Cruikshanks) [inline test/models/review-result.model.test.js:81]: I think something has gotten last in translation. This is not the pattern we use in our other model unit tests. This is what I would expect to see

```suggestion
        const result = await ReviewResultModel.query()
          .findById(testRecord.id)
          .withGraphFetched('reviewChargeElementResults')

        expect(result).to.be.instanceOf(ReviewResultModel)
        expect(result.id).to.equal(testRecord.id)

        expect(result.reviewChargeElementResults).to.be.an.array()
        expect(result.reviewChargeElementResults[0]).to.be.an.instanceOf(ReviewChargeElementResultModel)
        expect(result.reviewChargeElementResults).to.include(testReviewChargeElementResults[0])
        expect(result.reviewChargeElementResults).to.include(testReviewChargeElementResults[1])
```

File: test/models/review-result.model.test.js [hunk 11]
```diff
@@ -0,0 +1,128 @@
...
+      })
+    })
+
+    describe('when linking to review return result', () => {
```
Comment (Cruikshanks) [inline test/models/review-result.model.test.js:85]:
```suggestion
    describe('when linking to review return results', () => {
```

File: test/models/review-result.model.test.js [hunk 12]
```diff
@@ -0,0 +1,128 @@
...
+      })
+
+      it('can successfully run a related query', async () => {
+        const query = await ReviewResultModel.query()
+          .innerJoinRelated('reviewReturnResults')
+
+        expect(query).to.exist()
+      })
+
+      it('can eager load the ', async () => {
```
Comment (Cruikshanks) [inline test/models/review-result.model.test.js:106]:
```suggestion
      it('can eager load the review return results', async () => {
```

File: test/models/review-result.model.test.js [hunk 13]
```diff
@@ -0,0 +1,128 @@
...
+        expect(result[1].id).to.equal(testRecords[1].id)
+
+        expect(result[0].reviewReturnResults).to.be.an.array()
+        expect(result[1].reviewReturnResults).to.be.an.array()
+
+        expect(result[0].reviewReturnResults[0]).to.be.instanceOf(ReviewReturnResultModel)
+        expect(result[1].reviewReturnResults[0]).to.be.instanceOf(ReviewReturnResultModel)
+
+        expect(result[0].reviewReturnResults).to.include(testReviewReturnResult)
+        expect(result[1].reviewReturnResults).to.include(testReviewReturnResult)
```
Comment (Cruikshanks) [inline test/models/review-result.model.test.js:124]: Same as previous comment. This is the patten I was expecting to see

```suggestion
        const result = await ReviewResultModel.query()
          .findById(testRecord.id)
          .withGraphFetched('reviewReturnResults')

        expect(result).to.be.instanceOf(ReviewResultModel)
        expect(result.id).to.equal(testRecord.id)

        expect(result.reviewReturnResults).to.be.an.array()
        expect(result.reviewReturnResults[0]).to.be.an.instanceOf(ReviewReturnResultModel)
        expect(result.reviewReturnResults).to.include(testReviewReturnResults[0])
        expect(result.reviewReturnResults).to.include(testReviewReturnResults[1])
```

File: test/models/review-result.model.test.js [hunk 14]
```diff
@@ -0,0 +1,100 @@
...
+
+        expect(result).to.be.instanceOf(ReviewResultModel)
+        expect(result.id).to.equal(testRecord.id)
+
+        expect(result.reviewChargeElementResults).to.be.instanceOf(ReviewChargeElementResultModel)
+        expect(result.reviewChargeElementResults).to.equal(testReviewChargeElementResult)
+      })
+    })
+
+    describe('when linking to review return results', () => {
```
Comment (Cruikshanks) [inline test/models/review-result.model.test.js:70]:
```suggestion
    describe('when linking to review return result', () => {
```

File: test/models/review-result.model.test.js [hunk 15]
```diff
@@ -0,0 +1,100 @@
...
+    it('can successfully run a basic query', async () => {
+      const result = await ReviewResultModel.query().findById(testRecord.id)
+
+      expect(result).to.be.an.instanceOf(ReviewResultModel)
+      expect(result.id).to.equal(testRecord.id)
+    })
+  })
+
+  describe('Relationships', () => {
+    describe('when linking to review charge element results', () => {
```
Comment (Cruikshanks) [inline test/models/review-result.model.test.js:42]:
```suggestion
    describe('when linking to review charge element result', () => {
```

**Review by Cruikshanks — CHANGES_REQUESTED** [review]
Minor, _minor_ tweak to the test descriptions 😁

**Review by Cruikshanks — APPROVED** [review]

**Pull request merged**